File: backend/app/services/agent_security_event_store.py
"""
Agent Security Event Store
==========================
Event log for autonomous agent page scans.

Deliberately a separate store from `security_event_store`, which belongs to the
user-initiated "Scan Page" workflow. The two logs must never interleave: a user
reading their manual scan history should not see the agent's per-iteration
scans, and an aborted agent task must be attributable to its own task id.
"""

import uuid
from collections import deque
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional


class AgentSecurityEventStore:
    """In-memory log of agent page scans, capped at the most recent entries."""

    def __init__(self, max_size: int = 100):
        self._events = deque(maxlen=max_size)

    def add_event(
        self,
        task_id: str,
        url: str,
        allowed: bool,
        risk_level: str,
        summary_reason: str,
        blocked_sources: Optional[List[str]] = None,
    ) -> Dict[str, Any]:
        event = {
            "id": str(uuid.uuid4()),
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "task_id": task_id,
            "url": url,
            "allowed": allowed,
            "risk_level": risk_level,
            "summary_reason": summary_reason,
            "blocked_sources": blocked_sources or [],
            "origin": "agent_runtime",
        }
        self._events.append(event)
        return event

    def get_events(self, task_id: Optional[str] = None) -> List[Dict[str, Any]]:
        """Newest first, optionally narrowed to a single task."""
        events = list(reversed(self._events))
        if task_id:
            events = [event for event in events if event["task_id"] == task_id]
        return events

    def clear(self) -> None:
        self._events.clear()


agent_security_event_store = AgentSecurityEventStore()

File: backend-node/src/services/agentSecurityEventStore.ts
// Port of backend/app/services/agent_security_event_store.py.
//
// Deliberately a separate store from securityEventStore.ts, which belongs to
// the user-initiated "Scan Page" workflow. The two logs must never interleave:
// a user reading their manual scan history should not see the agent's
// per-iteration scans, and an aborted agent task must be attributable to its
// own task id.
import { randomUUID } from 'node:crypto'

export interface AgentSecurityEvent {
  id: string
  timestamp: string
  task_id: string
  url: string
  allowed: boolean
  risk_level: string
  summary_reason: string
  blocked_sources: string[]
  origin: 'agent_runtime'
}

const MAX_SIZE = 100

class AgentSecurityEventStore {
  private events: AgentSecurityEvent[] = []

  addEvent(
    taskId: string,
    url: string,
    allowed: boolean,
    riskLevel: string,
    summaryReason: string,
    blockedSources: readonly string[] = [],
  ): AgentSecurityEvent {
    const event: AgentSecurityEvent = {
      id: randomUUID(),
      timestamp: new Date().toISOString(),
      task_id: taskId,
      url,
      allowed,
      risk_level: riskLevel,
      summary_reason: summaryReason,
      blocked_sources: [...blockedSources],
      origin: 'agent_runtime',
    }
    this.events.push(event)
    if (this.events.length > MAX_SIZE) this.events.shift()
    return event
  }

  /** Newest first, optionally narrowed to a single task. */
  getEvents(taskId?: string): AgentSecurityEvent[] {
    const events = [...this.events].reverse()
    return taskId ? events.filter((event) => event.task_id === taskId) : events
  }

  clear(): void {
    this.events = []
  }
}

export const agentSecurityEventStore = new AgentSecurityEventStore()
